hotshot-query-service/src/fetching/provider/query_service.rs

```diff
@@ -2327,4 +2376,185 @@ mod test {
+        setup_test();
+
+        async fn run_test_helper<V: Versions>(port: u16, version: &str) {
+            let mut network = MockNetwork::<MockDataSource, V>::init().await;
+
+            let mut app = App::<_, Error>::with_state(ApiState::from(network.data_source()));
+
+            // Register availability APIs for two versions: v0 and v1
+            app.register_module(
+                "availability",
+                define_api(
+                    &Default::default(),
+                    StaticVersion::<0, 1> {},
+                    "0.0.1".parse().unwrap(),
+                )
+                .unwrap(),
+            )
+            .unwrap();
+
+            app.register_module(
+                "availability",
+                define_api(
+                    &Default::default(),
+                    StaticVersion::<0, 1> {},
+                    "1.0.0".parse().unwrap(),
+                )
+                .unwrap(),
+            )
+            .unwrap();
+
+            network.spawn(
+                "server",
+                app.serve(format!("0.0.0.0:{port}"), StaticVersion::<0, 1> {}),
+            );
+
+            let db = TmpDb::init().await;
+
+            let provider_url = format!("http://localhost:{port}/{version}")
+                .parse()
+                .expect("Invalid URL");
+
+            let provider = Provider::new(QueryServiceProvider::new(
+                provider_url,
+                StaticVersion::<0, 1> {},
+            ));
+
+            let ds = data_source(&db, &provider).await;
+            network.start().await;
+
+            let leaves = network.data_source().subscribe_leaves(1).await;
+            let leaves = leaves.take(5).collect::<Vec<_>>().await;
+            let test_leaf = &leaves[0];
+            let test_payload = &leaves[2];
+            let test_common = &leaves[3];
+
+            let mut fetches = vec![];
+            // Issue requests for missing data (these should initially remain unresolved):
+            fetches.push(ds.get_leaf(test_leaf.height() as usize).await.map(ignore));
+            fetches.push(ds.get_payload(test_payload.block_hash()).await.map(ignore));
+            fetches.push(
+                ds.get_vid_common(test_common.block_hash())
+                    .await
+                    .map(ignore),
+            );
+
+            // Even if we give data extra time to propagate, these requests will not resolve, since we
+            // didn't trigger any active fetches.
+            sleep(Duration::from_secs(1)).await;
+            for (i, fetch) in fetches.into_iter().enumerate() {
+                tracing::info!("checking fetch {i} is unresolved");
+                fetch.try_resolve().unwrap_err();
+            }
+
+            // Append the latest known leaf to the local store
+
+            ds.append(leaves.last().cloned().unwrap().into())
+                .await
+                .unwrap();
+
+            // This should now fetch from the provider
```

Comment:
Why exactly does this fetch now?

Reply:
we appended latest leaf before these calls, now these get calls would fetch from the provider and verify the whole leaf chain.